packages/realm-web/src/Fetcher.ts

```diff
@@ -0,0 +1,283 @@
+            } else {
+                // Throw an error with a message extracted from the body
+                throw await MongoDBRealmError.fromRequestAndResponse(
+                    request as Request<RequestBody>,
+                    response,
+                );
+            }
+        }
+    }
+
+    /**
+     * Fetch a network resource as an authenticated user and parse the result as extended JSON.
+     *
+     * @param request The request which should be sent to the server.
+     * @returns The response from the server, parsed as extended JSON.
+     */
+    public async fetchJSON<
+        RequestBody extends object = any,
+        ResponseBody extends object = any
+    >(request: AuthenticatedRequest<RequestBody>): Promise<ResponseBody> {
+        const { body } = request;
+        const serializedBody =
+            typeof body === "object" ? JSON.stringify(serialize(body)) : body;
+        const contentTypeHeaders = this.buildJsonHeader(body);
+        const response = await this.fetch<RequestBody>({
+            ...request,
+            body: serializedBody,
+            headers: {
+                Accept: "application/json",
+                ...contentTypeHeaders,
+                ...request.headers,
+            },
+        });
+        const contentType = response.headers.get("content-type");
+        if (contentType?.startsWith("application/json")) {
+            const responseBody = await response.json();
+            return deserialize<ResponseBody>(responseBody);
+        } else if (contentType === null) {
+            return (null as unknown) as ResponseBody;
+        } else {
+            throw new Error(`Expected JSON response, got "${contentType}"`);
+        }
+    }
+
+    /**
+     * @returns A promise of the app URL, with the app location resolved.
+     */
+    public async getLocationUrl() {
+        if (!this.location) {
+            const path = routes.api().app(this.appId).location().path;
+            this.location = this.fetchJSON({
+                method: "GET",
+                url: this.baseUrl + path,
+                tokenType: "none",
+            });
+        }
+        // Return the hostname as the location URL.
+        const { hostname } = await this.location;
+        if (typeof hostname !== "string") {
+            throw new Error("Expected response to contain a 'hostname'");
+        }
+        return hostname;
+    }
+
+    /**
+     * @returns A promise of the app URL, with the app location resolved.
+     */
+    public getAppPath() {
+        return routes.api().app(this.appId);
+    }
+
+    /**
+     * @param user An optional user to generate the header for.
+     * @param tokenType The type of token (access or refresh).
+     * @returns An object containing with the users access token as authorization header or undefined if no user is given.
+     */
+    private buildAuthorizationHeader(user: User | null, tokenType: TokenType) {
+        if (user && tokenType === "access") {
```

Comment:
I think it's a personal preference. I don't like when JS-code is explicit about returning `undefined`.

Reply:
While I agree for functions that logically return `void`, in this case I find it more confusing because this function returns `undefined` as a value in itself, rather than representing that this function doesn't have any return value (subtle distinction, I know). When I read it, my first thought was "well what happens when `!user`?", and I had to remember that it uses the default of returning `undefined`, rather than being a buggy forgotten case. 

I'll leave it to your judgement since I'm just passing through, but that is my 2¢.